repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeRoutingServlet.scala

```diff
@@ -0,0 +1,362 @@
+
+  def doRoute(servletRequest: HttpServletRequestWrapper, servletResponse: HttpServletResponse): Unit = {
+    val attributeDestinations =
+      Option(servletRequest.getAttribute(CommonRequestAttributes.DESTINATIONS))
+        .getOrElse(List.empty[RouteDestination]).asInstanceOf[List[RouteDestination]]
+    val defaultDestinations = {
+      defaultDestination.get() match {
+        case Some(defDest) =>
+          List[RouteDestination] {
+            new RouteDestination(defDest.getId, servletRequest.getRequestURI, -1)
+          }
+        case _ => List.empty[RouteDestination]
+      }
+    }
+    (attributeDestinations ++ defaultDestinations)
+      .sortWith(_.compareTo(_) > 0) // Highest Quality first
+      .headOption
+      .foreach { routeDest =>
+        localNode.get() match {
+          case Some(node) => doDestination(servletRequest, servletResponse, routeDest, node)
+          case _ =>
+            logger.error("Invalid local node definition")
+            servletResponse.setStatus(HttpServletResponse.SC_SERVICE_UNAVAILABLE)
+        }
+      }
+  }
+
+  def doDestination(servletRequest: HttpServletRequestWrapper, servletResponse: HttpServletResponse, routeDest: RouteDestination, node: Node): Unit = {
+    destinations.get().get(routeDest.getDestinationId) match {
+      case Some(destination) =>
+        localNode.get() match {
+          case Some(localhost) =>
+            getDestinationLocation(servletRequest, destination, localhost, routeDest.getUri, servletRequest) match {
+              case Some(locationUrl) =>
+                // According to the Java 6 javadocs the routeDestination passed into getContext:
+                // "The given path [routeDestination] must begin with /, is interpreted relative to the server's document root
+                // and is matched against the context roots of other web applications hosted on this container."
+                Option(getServletContext.getContext(locationUrl.getPath)).foreach { targetContext =>
+                  // Capture this for Location header processing
+                  val locationUrlPath = locationUrl.getPath
+                  val targetContextPath = targetContext.getContextPath
+                  val dispatchPath = if (locationUrlPath.startsWith(targetContextPath)) locationUrlPath.substring(targetContextPath.length) else locationUrlPath
+                  Option(targetContext.getRequestDispatcher(dispatchPath)) match {
+                    case Some(dispatcher) =>
+                      doDispatch(
+                        servletRequest,
+                        servletResponse,
+                        locationUrl,
+                        dispatchPath,
+                        targetContext,
+                        routeDest,
+                        dispatcher,
+                        destination)
+                    case _ =>
+                      logger.warn(s"No Request Dispatcher found in Servlet Context for $dispatchPath; not forwarding")
```

Comment:
I'm not certain warn is a high enough log. I could be convinced otherwise though.